src/lambdaisland/embedkit.clj

```diff
@@ -120,40 +138,62 @@
                                                     "password" password}
                                       :content-type :json
                                       :as :json})))]
-    (map->Connection {:client client
-                      :endpoint endpoint
-                      :token token
-                      :cache (atom {})
-                      :secret-key secret-key
-                      :middleware middleware})))
+    (map->Connection (assoc
+                       conn-opts
+                       :client client
+                       :endpoint endpoint
+                       :token token
+                       :cache (atom {})
+                       :secret-key secret-key
+                       :middleware middleware))))
 
 (defn- do-request [req]
   (vary-meta (http/request req)
              assoc
              ::request req))
 
+(defmacro with-refresh-auth [conn & body]
+  `((fn retry# [conn# retries#]
+      (Thread/sleep (* 1000 retries# retries#)) ; backoff
+      (try
+        ~@body
+        (catch clojure.lang.ExceptionInfo e#
+          (if (and (= 401 (:status (ex-data e#)))
+                   (atom? conn#)
+                   (< 4 retries#))
+            (do
+              (reset! conn# (connect @conn#))
+              (retry# conn# (inc retries#)))
+            (throw e#)))))
+    ~conn
+    0))
```

Comment:
Pretty standard retry loop. At first iteration `retries#` is 0, so we don't wait. Then we wait 1 second, 4 seconds, 8 seconds, 16 seconds, before giving up.

Doesn't do anything special (i.e. preserves existing behavior) unless
- the connection is an atom
- we get a 401